pam/integration-tests/ssh_test.go

```diff
@@ -0,0 +1,557 @@
+		t.Name())
+	if daemonize {
+		// When in daemon mode SSH doesn't show debug infos, so let's
+		// mange this manually.
+		sshdConnectCommand += "&& env | sort | sed 's/^/  /'"
+	}
+	if interactiveShell {
+		sshdConnectCommand = "/bin/sh"
+	}
+
+	userHome := t.TempDir()
+	sshdPort := startSSHd(t, hostKey, sshdConnectCommand, []string{
+		fmt.Sprintf("HOME=%s", userHome),
+		fmt.Sprintf("LD_PRELOAD=%s", preloadLibrary),
+		fmt.Sprintf("AUTHD_TEST_SSH_USER=%s", user),
+		fmt.Sprintf("AUTHD_TEST_SSH_HOME=%s", userHome),
+		fmt.Sprintf("AUTHD_TEST_SSH_PAM_SERVICE=%s", serviceFile),
+	}, daemonize)
+
+	return sshdPort, userHome
+}
+
+func sshdCommand(t *testing.T, port, hostKey, forcedCommand string, env []string, daemonize bool) (*exec.Cmd, string, string) {
+	t.Helper()
+
+	logFile := ""
+	pidFile := ""
+	runModeArgs := []string{"-ddd"}
+
+	if daemonize {
+		pidFile = filepath.Join(t.TempDir(), "sshd.pid")
+		logFile = filepath.Join(t.TempDir(), "sshd-daemon.log")
+		saveArtifactsForDebugOnCleanup(t, []string{logFile})
+
+		runModeArgs = []string{
+			"-E", logFile,
+			"-o", "PidFile=" + pidFile,
+			"-o", "LogLevel=DEBUG3",
+		}
+	}
+
+	// #nosec:G204 - we control the command arguments in tests
+	sshd := exec.Command("/usr/sbin/sshd",
+		"-f", os.DevNull,
+		"-p", port,
+		"-h", hostKey,
+		"-o", "UsePAM=yes",
+		"-o", "KbdInteractiveAuthentication=yes",
+		"-o", "AuthenticationMethods=keyboard-interactive",
+		"-o", "IgnoreUserKnownHosts=yes",
+		"-o", "AuthorizedKeysFile=none",
+		"-o", "PermitUserEnvironment=no",
+		"-o", "PermitUserRC=no",
+		"-o", "ClientAliveInterval=300",
+		"-o", "ClientAliveCountMax=3",
+		"-o", "ForceCommand="+forcedCommand,
+	)
+	sshd.Args = append(sshd.Args, runModeArgs...)
+	sshd.Env = append(sshd.Env, env...)
+	sshd.Env = testutils.AppendCovEnv(sshd.Env)
+
+	return sshd, pidFile, logFile
+}
+
+// safeBuffer use locked read/writes on bytes.Buffer to avoid races, we only
+// override the methods we require in the tests.
+type safeBuffer struct {
+	bytes.Buffer
+	mu sync.RWMutex
+}
+
+func (sb *safeBuffer) Write(p []byte) (n int, err error) {
+	sb.mu.Lock()
+	defer sb.mu.Unlock()
+
+	return sb.Buffer.Write(p)
+}
+
+func (sb *safeBuffer) ReadFrom(r io.Reader) (n int64, err error) {
+	sb.mu.Lock()
+	defer sb.mu.Unlock()
+
+	return sb.Buffer.ReadFrom(r)
+}
+
+func (sb *safeBuffer) String() string {
+	sb.mu.RLock()
+	defer sb.mu.RUnlock()
+
+	return sb.Buffer.String()
+}
+
+func startSSHd(t *testing.T, hostKey, forcedCommand string, env []string, daemonize bool) string {
+	t.Helper()
+
+	// We use this to easily find a free port we can use, without going random
+	server := httptest.NewServer(http.HandlerFunc(nil))
+	url, err := url.Parse(server.URL)
+	require.NoError(t, err, "Setup: Impossible to find a valid port to use")
+	sshdPort := url.Port()
+	server.Close()
+
+	sshd, sshdPidFile, sshdLogFile := sshdCommand(t, sshdPort, hostKey, forcedCommand, env, daemonize)
+	sshdStderr := safeBuffer{}
```

Comment:
I had put 

```
// safeBuffer use locked read/writes on bytes.Buffer to avoid races, we only
// override the methods we require in the tests.
```
Want me to elaborate more?

Reply:
Yes please. I've seen that comment but it's not clear to me *why* we need to care about races here.

Reply:
I've elaborated it more in 01807f3a1e2b98ee25dcef02849564d4b3478c15 let me know if it's clearer now.

Reply:
Yes, that makes the intention clearer, thanks! However, as far as I can tell, the cleanup functions registered via `t.Cleanup` are not started in separate goroutines, see https://github.com/golang/go/blob/go1.23.1/src/testing/testing.go#L1354-L1354. Am I missing something?

Reply:
No, `t.Cleanup` is started on the main test goroutine, but we do start another goroutine in our cleanup function and that was racing with the exec.Command side.

Disabling that, when running with `-race` we easily end up with:

```
WARNING: DATA RACE
Read at 0x00c000129800 by goroutine 9:
  runtime.slicebytetostring()
      /snap/go/10730/src/runtime/string.go:91 +0x0
  bytes.(*Buffer).String()
      /snap/go/10730/src/bytes/buffer.go:71 +0xc18
  github.com/ubuntu/authd/pam/integration-tests_test.(*safeBuffer).String()
      /home/marco/Dev/authd/pam/integration-tests/ssh_test.go:420 +0xbb5
  github.com/ubuntu/authd/pam/integration-tests_test.startSSHd()
      /home/marco/Dev/authd/pam/integration-tests/ssh_test.go:554 +0xc63
  github.com/ubuntu/authd/pam/integration-tests_test.startSSHdForTest()
      /home/marco/Dev/authd/pam/integration-tests/ssh_test.go:339 +0x52f
  github.com/ubuntu/authd/pam/integration-tests_test.testSSHAuthenticate()
      /home/marco/Dev/authd/pam/integration-tests/ssh_test.go:80 +0xa18
  github.com/ubuntu/authd/pam/integration-tests_test.TestSSHAuthenticate()
      /home/marco/Dev/authd/pam/integration-tests/ssh_test.go:35 +0x88
  testing.tRunner()
      /snap/go/10730/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /snap/go/10730/src/testing/testing.go:1743 +0x44

Previous write at 0x00c000129800 by goroutine 30:
  runtime.slicecopy()
      /snap/go/10730/src/runtime/slice.go:355 +0x0
  bytes.growSlice()
      /snap/go/10730/src/bytes/buffer.go:250 +0x144
  bytes.(*Buffer).grow()
      /snap/go/10730/src/bytes/buffer.go:151 +0x2b7
  bytes.(*Buffer).ReadFrom()
      /snap/go/10730/src/bytes/buffer.go:209 +0x65
  github.com/ubuntu/authd/pam/integration-tests_test.(*safeBuffer).ReadFrom()
      /home/marco/Dev/authd/pam/integration-tests/ssh_test.go:413 +0x3c
  io.copyBuffer()
      /snap/go/10730/src/io/io.go:415 +0x21e
  io.Copy()
      /snap/go/10730/src/io/io.go:388 +0x66
  os.genericWriteTo()
      /snap/go/10730/src/os/file.go:275 +0x12
  os.(*File).WriteTo()
      /snap/go/10730/src/os/file.go:253 +0xd4
  io.copyBuffer()
      /snap/go/10730/src/io/io.go:411 +0xd3
  io.Copy()
      /snap/go/10730/src/io/io.go:388 +0x5a
  os/exec.(*Cmd).writerDescriptor.func1()
      /snap/go/10730/src/os/exec/exec.go:580 +0x37
  os/exec.(*Cmd).Start.func2()
      /snap/go/10730/src/os/exec/exec.go:733 +0x3c
  os/exec.(*Cmd).Start.gowrap1()
      /snap/go/10730/src/os/exec/exec.go:745 +0x41

Goroutine 9 (running) created at:
  testing.(*T).Run()
      /snap/go/10730/src/testing/testing.go:1743 +0x825
  testing.runTests.func1()
      /snap/go/10730/src/testing/testing.go:2168 +0x85
  testing.tRunner()
      /snap/go/10730/src/testing/testing.go:1690 +0x226
  testing.runTests()
      /snap/go/10730/src/testing/testing.go:2166 +0x8be
  testing.(*M).Run()
      /snap/go/10730/src/testing/testing.go:2034 +0xf17
  github.com/ubuntu/authd/pam/integration-tests_test.TestMain()
      /home/marco/Dev/authd/pam/integration-tests/cli-integration_test.go:27 +0x244
  main.main()
      _testmain.go:69 +0x171

Goroutine 30 (running) created at:
  os/exec.(*Cmd).Start()
      /snap/go/10730/src/os/exec/exec.go:732 +0x11b9
  github.com/ubuntu/authd/pam/integration-tests_test.startSSHd()
      /home/marco/Dev/authd/pam/integration-tests/ssh_test.go:442 +0x551
  github.com/ubuntu/authd/pam/integration-tests_test.startSSHdForTest()
      /home/marco/Dev/authd/pam/integration-tests/ssh_test.go:339 +0x52f
  github.com/ubuntu/authd/pam/integration-tests_test.testSSHAuthenticate()
      /home/marco/Dev/authd/pam/integration-tests/ssh_test.go:80 +0xa18
  github.com/ubuntu/authd/pam/integration-tests_test.TestSSHAuthenticate()
      /home/marco/Dev/authd/pam/integration-tests/ssh_test.go:35 +0x88
  testing.tRunner()
      /snap/go/10730/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /snap/go/10730/src/testing/testing.go:1743 +0x44
```

Reply:
I hope this is resolved too :)